File: Algorithm/Kadane’s Algorithm/leetcode/53. Maximum Subarray/Claude/Maximum-Subarray.py
# PythonでのMaximum Subarrayの実装を、LeetCodeの形式に合わせて型安全かつパフォーマンス最適化して作成します。Python実装のポイントを詳しく説明します：

# ## 🎯 **LeetCode最適化のポイント**

# ### **1. 型安全性（Pylance対応）**
# - `typing.List`を使用して厳密な型指定
# - すべての変数に型アノテーション
# - 戻り値の型も明示的に指定

# ### **2. パフォーマンス最適化**
# ```python
# # ❌ 遅い実装
# current_sum = max(nums[i], current_sum + nums[i])

# # ✅ 高速実装
# current_sum = current_sum + nums[i] if current_sum > 0 else nums[i]
# ```

# ### **3. メモリ効率化**
# - `O(1)`空間計算量を維持
# - 不要な配列やリストの作成を回避
# - インプレース処理を重視

# ## 🚀 **推奨実装順位**

# 1. **`SolutionFinal.maxSubArray`** - LeetCode提出用最推奨
# 2. **`maxSubArrayOptimized`** - 可読性とパフォーマンスのバランス
# 3. **`maxSubArrayIterative`** - Pythonic実装

# ## ⚡ **実行時間とメモリ使用量の考慮事項**

# ### **避けるべきパターン**
# ```python
# # ❌ 関数呼び出しオーバーヘッド
# max_sum = max(max_sum, current_sum)

# # ✅ 条件分岐で最適化
# if current_sum > max_sum:
#     max_sum = current_sum
# ```

# ### **Python特有の最適化**
# - スライス操作の回避（`nums[1:]`など）
# - リスト内包表記よりもシンプルなループ
# - `sys.maxsize`の使用で無限大を表現

# ## 📊 **各実装の特徴**

# | 実装 | 時間計算量 | 空間計算量 | LeetCode性能 | 用途 |
# |------|------------|------------|--------------|------|
# | `SolutionFinal` | O(n) | O(1) | 🟢 最高 | **提出用推奨** |
# | `maxSubArrayOptimized` | O(n) | O(1) | 🟢 優秀 | 学習・理解用 |
# | `maxSubArrayDivideConquer` | O(n log n) | O(log n) | 🟡 普通 | 面接・学習用 |
# | `maxSubArrayDP` | O(n) | O(n) | 🟡 普通 | DP学習用 |

# **LeetCode提出には`SolutionFinal`クラスの実装を使用してください！** 最も効率的で、実行時間とメモリ使用量の両方で最適化されています。

from typing import List
import sys


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        """
        最大部分配列の和を求める（Kadane's Algorithm）
        
        Args:
            nums: 整数のリスト
            
        Returns:
            最大部分配列の和
            
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        # 現在の部分配列の和を追跡
        current_sum: int = nums[0]
        # これまでに見つけた最大の部分配列の和
        max_sum: int = nums[0]
        
        # 配列の2番目の要素から開始
        for i in range(1, len(nums)):
            # 現在の要素から新しく開始するか、既存の部分配列に追加するかを選択
            current_sum = max(nums[i], current_sum + nums[i])
            
            # 最大和を更新
            max_sum = max(max_sum, current_sum)
        
        return max_sum
    
    def maxSubArrayOptimized(self, nums: List[int]) -> int:
        """
        最大部分配列の和を求める（LeetCode最適化版）
        
        Args:
            nums: 整数のリスト
            
        Returns:
            最大部分配列の和
            
        Time Complexity: O(n)
        Space Complexity: O(1)
        
        Note: max()関数の呼び出し回数を減らしてパフォーマンス向上
        """
        max_sum: int = nums[0]
        current_sum: int = nums[0]
        
        for i in range(1, len(nums)):
            # 条件分岐でmax()を回避してパフォーマンス向上
            if current_sum > 0:
                current_sum += nums[i]
            else:
                current_sum = nums[i]
            
            # max()を回避してif文を使用
            if current_sum > max_sum:
                max_sum = current_sum
        
        return max_sum
    
    def maxSubArrayDivideConquer(self, nums: List[int]) -> int:
        """
        最大部分配列の和を求める（分割統治法）
        
        Args:
            nums: 整数のリスト
            
        Returns:
            最大部分配列の和
            
        Time Complexity: O(n log n)
        Space Complexity: O(log n)
        """
        def divide_conquer(left: int, right: int) -> int:
            """
            分割統治法のヘルパー関数
            
            Args:
                left: 左端のインデックス
                right: 右端のインデックス
                
            Returns:
                指定範囲での最大部分配列の和
            """
            # ベースケース: 要素が1つの場合
            if left == right:
                return nums[left]
            
            # 中点を計算
            mid: int = left + (right - left) // 2
            
            # 左半分の最大部分配列の和
            left_max: int = divide_conquer(left, mid)
            
            # 右半分の最大部分配列の和
            right_max: int = divide_conquer(mid + 1, right)
            
            # 中点をまたぐ最大部分配列の和を計算
            left_sum: int = -sys.maxsize
            total: int = 0
            for i in range(mid, left - 1, -1):
                total += nums[i]
                left_sum = max(left_sum, total)
            
            right_sum: int = -sys.maxsize
            total = 0
            for i in range(mid + 1, right + 1):
                total += nums[i]
                right_sum = max(right_sum, total)
            
            cross_sum: int = left_sum + right_sum
            
            # 3つの候補の中から最大値を返す
            return max(left_max, right_max, cross_sum)
        
        return divide_conquer(0, len(nums) - 1)
    
    def maxSubArrayWithIndices(self, nums: List[int]) -> tuple[int, int, int]:
        """
        最大部分配列の和と開始・終了インデックスを返す
        
        Args:
            nums: 整数のリスト
            
        Returns:
            (最大和, 開始インデックス, 終了インデックス)のタプル
            
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        max_sum: int = nums[0]
        current_sum: int = nums[0]
        start: int = 0
        end: int = 0
        temp_start: int = 0
        
        for i in range(1, len(nums)):
            if current_sum < 0:
                current_sum = nums[i]
                temp_start = i
            else:
                current_sum += nums[i]
            
            if current_sum > max_sum:
                max_sum = current_sum
                start = temp_start
                end = i
        
        return max_sum, start, end
    
    def maxSubArrayDP(self, nums: List[int]) -> int:
        """
        最大部分配列の和を求める（動的プログラミング版）
        メモリ使用量は多いが理解しやすい実装
        
        Args:
            nums: 整数のリスト
            
        Returns:
            最大部分配列の和
            
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        n: int = len(nums)
        # dp[i]は位置iで終わる最大部分配列の和
        dp: List[int] = [0] * n
        dp[0] = nums[0]
        max_sum: int = nums[0]
        
        for i in range(1, n):
            # 前の部分配列に追加するか、新しく開始するかを選択
            dp[i] = max(nums[i], dp[i-1] + nums[i])
            max_sum = max(max_sum, dp[i])
        
        return max_sum
    
    def maxSubArrayIterative(self, nums: List[int]) -> int:
        """
        最大部分配列の和を求める（イテレーティブ版、メモリ最適化）
        LeetCode提出用の推奨実装
        
        Args:
            nums: 整数のリスト
            
        Returns:
            最大部分配列の和
            
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        # max_ending_here: int = max_so_far: int = nums[0]
        
        # for num in nums[1:]:
        #     # より効率的な処理: スライス操作を回避
        #     max_ending_here = max_ending_here + num if max_ending_here > 0 else num
        #     max_so_far = max_so_far if max_so_far > max_ending_here else max_ending_here
        
        # return max_so_far
        max_ending_here: int = nums[0]
        max_so_far: int = nums[0]
        
        for i in range(1, len(nums)):
            # より効率的な処理: スライス操作を回避し、インデックスアクセスを使用
            if max_ending_here > 0:
                max_ending_here = max_ending_here + nums[i]
            else:
                max_ending_here = nums[i]
            
            if max_ending_here > max_so_far:
                max_so_far = max_ending_here
        
        return max_so_far


# 使用例とパフォーマンス比較用の関数
def performance_comparison() -> None:
    """
    各実装のパフォーマンス比較（デモ用）
    実際のLeetCode提出時は不要
    """
    import time
    import random
    
    # テストデータ生成
    test_data: List[int] = [random.randint(-100, 100) for _ in range(10000)]
    solution = Solution()
    
    methods = [
        ("Kadane's Algorithm", solution.maxSubArray),
        ("Optimized Version", solution.maxSubArrayOptimized),
        ("Iterative Version", solution.maxSubArrayIterative),
        ("DP Version", solution.maxSubArrayDP),
    ]
    
    print("Performance Comparison (10,000 elements):")
    print("-" * 50)
    
    for name, method in methods:
        start_time = time.perf_counter()
        result = method(test_data)
        end_time = time.perf_counter()
        
        print(f"{name:20}: {end_time - start_time:.6f}s, Result: {result}")


# LeetCode提出用の最終推奨実装
class SolutionFinal:
    """
    LeetCode提出用の最終実装（最も効率的）
    """
    def maxSubArray(self, nums: List[int]) -> int:
        """
        最大部分配列の和を求める（最終最適化版）
        
        Args:
            nums: 整数のリスト (1 <= len(nums) <= 10^5, -10^4 <= nums[i] <= 10^4)
            
        Returns:
            最大部分配列の和
            
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        max_sum = current_sum = nums[0]
        
        for i in range(1, len(nums)):
            current_sum = current_sum + nums[i] if current_sum > 0 else nums[i]
            if current_sum > max_sum:
                max_sum = current_sum
        
        return max_sum

File: Algorithm/Kadane’s Algorithm/leetcode/53. Maximum Subarray/Claude/Maximum-Subarray.js
// この問題は「Maximum Subarray Problem」で、Kadane's Algorithmを使って効率的に解くことができます。O(n)時間計算量とO(1)空間計算量で実装します。さらに、分割統治法（Divide and Conquer）でのアプローチも実装します：**解説：**

// 1. **Kadane's Algorithm（推奨）**：
//    - 各位置で「現在の要素から新しく開始する」か「既存の部分配列に追加する」かを選択
//    - より効率的でLeetCodeでの実行時間とメモリ使用量が最適
//    - 時間計算量：O(n)、空間計算量：O(1)

// 2. **分割統治法**：
//    - 配列を左半分、右半分、中点をまたぐ部分に分けて考える
//    - より理論的なアプローチだが、実行時間とメモリ使用量は劣る
//    - 時間計算量：O(n log n)、空間計算量：O(log n)

// **LeetCode提出用には最初の`maxSubArray`関数を使用することを推奨します。**

// 実装のポイント：
// - 単一ループで効率的に処理
// - 不要な変数や配列を使わずメモリ使用量を最小化
// - `Math.max`を使って可読性を保ちながら最適化

/**
 * 最大部分配列の和を求める関数（Kadane's Algorithm）
 * @param {number[]} nums - 整数配列
 * @return {number} - 最大部分配列の和
 * 
 * 時間計算量: O(n)
 * 空間計算量: O(1)
 */
var maxSubArray = function(nums) {
    // 現在の部分配列の和を追跡
    let currentSum = nums[0];
    // これまでに見つけた最大の部分配列の和
    let maxSum = nums[0];
    
    // 配列の2番目の要素から開始
    for (let i = 1; i < nums.length; i++) {
        // 現在の要素から新しく開始するか、既存の部分配列に追加するかを選択
        // より大きい値を選ぶ
        currentSum = Math.max(nums[i], currentSum + nums[i]);
        
        // 最大和を更新
        maxSum = Math.max(maxSum, currentSum);
    }
    
    return maxSum;
};

/**
 * 最大部分配列の和を求める関数（分割統治法）
 * @param {number[]} nums - 整数配列
 * @return {number} - 最大部分配列の和
 * 
 * 時間計算量: O(n log n)
 * 空間計算量: O(log n) - 再帰呼び出しのスタック
 */
var maxSubArrayDivideConquer = function(nums) {
    /**
     * 分割統治法のヘルパー関数
     * @param {number[]} nums - 配列
     * @param {number} left - 左端のインデックス
     * @param {number} right - 右端のインデックス
     * @return {number} - 指定範囲での最大部分配列の和
     */
    function divideConquer(nums, left, right) {
        // ベースケース: 要素が1つの場合
        if (left === right) {
            return nums[left];
        }
        
        // 中点を計算
        const mid = Math.floor((left + right) / 2);
        
        // 左半分の最大部分配列の和
        const leftMax = divideConquer(nums, left, mid);
        
        // 右半分の最大部分配列の和
        const rightMax = divideConquer(nums, mid + 1, right);
        
        // 中点をまたぐ最大部分配列の和を計算
        let leftSum = Number.NEGATIVE_INFINITY;
        let sum = 0;
        for (let i = mid; i >= left; i--) {
            sum += nums[i];
            leftSum = Math.max(leftSum, sum);
        }
        
        let rightSum = Number.NEGATIVE_INFINITY;
        sum = 0;
        for (let i = mid + 1; i <= right; i++) {
            sum += nums[i];
            rightSum = Math.max(rightSum, sum);
        }
        
        const crossSum = leftSum + rightSum;
        
        // 3つの候補の中から最大値を返す
        return Math.max(leftMax, rightMax, crossSum);
    }
    
    return divideConquer(nums, 0, nums.length - 1);
};